src/mlpack/core/dists/discrete_distribution.cpp

```diff
@@ -44,33 +51,45 @@ arma::vec DiscreteDistribution::Random() const
  */
 void DiscreteDistribution::Train(const arma::mat& observations)
 {
-  // Clear old probabilities.
-  probabilities.zeros();
+  // Make sure the observations have same dimension with the probabilities
+  if(observations.n_rows != probabilities.size())
+  {
+    Log::Debug << "the obversation must has the same dimension with the probabilities"
+        << "the observation's dimension is" << observations.n_cols << "but the dimension of "
+        << "probabilities is" << probabilities.size() << std::endl;
+  }
+  // Get the dimension size of the distribution
+  const size_t dimensions = probabilities.size();
 
-  // Add the probability of each observation.  The addition of 0.5 to the
-  // observation is to turn the default flooring operation of the size_t cast
-  // into a rounding operation.
-  for (size_t i = 0; i < observations.n_cols; i++)
+  // Iterate all the probabilities in each dimension
+  for (size_t i=0; i < dimensions; i++)
   {
-    const size_t obs = size_t(observations(0, i) + 0.5);
+    // Clear the old probabilities
+    probabilities[i].zeros();
+    for (size_t r=0; r < observations.n_cols; r++)
+      {
+      // Add the probability of each observation.  The addition of 0.5 to the
+      // observation is to turn the default flooring operation of the size_t cast
+      // into a rounding observation.
+      const size_t obs = size_t(observations(i, r) + 0.5);
 
-    // Ensure that the observation is within the bounds.
-    if (obs >= probabilities.n_elem)
-    {
-      Log::Debug << "DiscreteDistribution::Train(): observation " << i
-          << " (" << obs << ") is invalid; observation must be in [0, "
-          << probabilities.n_elem << "] for this distribution." << std::endl;
-    }
+      // Ensure that the observation is within the bounds.
+      if (obs >= probabilities[i].n_elem)
+      {
+        Log::Debug << "DiscreteDistribution::Train(): observation " << i
+            << " (" << obs << ") is invalid; observation must be in [0, "
+            << probabilities[i].n_elem << "] for this distribution." << std::endl;       
+      }
+      probabilities[i][obs]++;
+      }
 
-    probabilities(obs)++;
+    // Now normailze the distribution.
+    double sum = accu(probabilities[i]);
+    if (sum > 0)
+      probabilities[i] /= sum;
+    else
+      probabilities[i].fill(1.0 / probabilities[i].n_elem); // Force normalization.
```

Comment:
Hi @rcurtin , hmmm, I'm not sure if I got the truly meaning of invert. I just write a demon so no need to review it again. :)  NOTE, I use another for loop to loop over each dimension of probabilities to normailze it and that's the reason why I use **dimensions** as outer loop's index in the first place.

```
  // Iterate observation in each dimension
  for (size_t r=0; r < observations.n_rows; r++)
  {
    // Iterate all the probabilities in each dimension
    for (size_t i=0; i < dimensions; i++)
    {
      // Clear the old probabilities
      probabilities[i].zeros();

      // Add the probability of each observation.  The addition of 0.5 to the
      // observation is to turn the default flooring operation of the size_t cast
      // into a rounding observation.
      const size_t obs = size_t(observations(i, r) + 0.5);
      
      // Ensure that the observation is within the bounds.
      if (obs >= probabilities[i].n_elem)
      {
        Log::Debug << "DiscreteDistribution::Train(): observation " << i
            << " (" << obs << ") is invalid; observation must be in [0, "
            << probabilities[i].n_elem << "] for this distribution." << std::endl;       
      }
      probabilities[i][obs]++;
    }
  }

  // Now normailze the distribution.
  for (size_t i=0; i < dimensions; i++)
  {
    double sum = accu(probabilities[i]);
    if (sum > 0)
      probabilities[i] /= sum;
    else
      probabilities[i].fill(1.0 / probabilities[i].n_elem); // Force normalization.
  }
```